File: src/channels/message-queue.ts
/**
 * MessageQueue — data hub for the Channel Adapter.
 *
 * Two async queues decouple message ingestion from delivery:
 *   inbound:  external channel → agent (produced by Channel, consumed by GatewayAdapter)
 *   outbound: agent → external channel (produced by GatewayAdapter, consumed by ChannelManager)
 */
import { createLogger } from "../logger";
import type { InboundMessage, OutboundMessage } from "../types";

const log = createLogger("channel:mq");

export class MessageQueue {
  private inboundBuffer: InboundMessage[] = [];
  private outboundBuffer: OutboundMessage[] = [];
  private inboundWaiters: Array<(msg: InboundMessage | null) => void> = [];
  private outboundWaiters: Array<(msg: OutboundMessage | null) => void> = [];
  private _stopped = false;

  publishInbound(msg: InboundMessage): void {
    if (this._stopped) return;
    log.debug("Inbound published", { channel: msg.channelType, peer: msg.peerId });
    if (this.inboundWaiters.length > 0) {
      this.inboundWaiters.shift()!(msg);
    } else {
      this.inboundBuffer.push(msg);
    }
  }

  publishOutbound(msg: OutboundMessage): void {
    if (this._stopped) return;
    log.debug("Outbound published", { channel: msg.channelType, peer: msg.peerId });
    if (this.outboundWaiters.length > 0) {
      this.outboundWaiters.shift()!(msg);
    } else {
      this.outboundBuffer.push(msg);
    }
  }

  consumeInbound(): Promise<InboundMessage | null> {
    if (this._stopped) return Promise.resolve(null);
    if (this.inboundBuffer.length > 0) {
      return Promise.resolve(this.inboundBuffer.shift()!);
    }
    return new Promise((resolve) => {
      this.inboundWaiters.push(resolve);
    });
  }

  consumeOutbound(): Promise<OutboundMessage | null> {
    if (this._stopped) return Promise.resolve(null);
    if (this.outboundBuffer.length > 0) {
      return Promise.resolve(this.outboundBuffer.shift()!);
    }
    return new Promise((resolve) => {
      this.outboundWaiters.push(resolve);
    });
  }

  stop(): void {
    this._stopped = true;
    for (const resolve of this.inboundWaiters) resolve(null);
    for (const resolve of this.outboundWaiters) resolve(null);
    this.inboundWaiters = [];
    this.outboundWaiters = [];
    log.info("MessageQueue stopped");
  }

  get inboundSize(): number {
    return this.inboundBuffer.length;
  }

  get outboundSize(): number {
    return this.outboundBuffer.length;
  }
}

let _mq: MessageQueue | null = null;

export function getMessageQueue(): MessageQueue {
  if (!_mq) _mq = new MessageQueue();
  return _mq;
}

File: src/channels/message_queue.py
"""Dual async message queue (inbound/outbound), matching TS channels/message-queue.ts."""

from __future__ import annotations

import asyncio

from src.types import IncomingMessage, OutgoingMessage
from src.utils.logger import create_logger

log = create_logger("channels:queue")


class MessageQueue:
    def __init__(self) -> None:
        self._inbound: asyncio.Queue[IncomingMessage | None] = asyncio.Queue()
        self._outbound: asyncio.Queue[OutgoingMessage | None] = asyncio.Queue()
        self._stopped = False

    async def publish_inbound(self, msg: IncomingMessage) -> None:
        await self._inbound.put(msg)

    async def publish_outbound(self, msg: OutgoingMessage) -> None:
        await self._outbound.put(msg)

    async def consume_inbound(self) -> IncomingMessage | None:
        return await self._inbound.get()

    async def consume_outbound(self) -> OutgoingMessage | None:
        return await self._outbound.get()

    def stop(self) -> None:
        self._stopped = True
        # Unblock consumers
        self._inbound.put_nowait(None)
        self._outbound.put_nowait(None)

    @property
    def inbound_size(self) -> int:
        return self._inbound.qsize()

    @property
    def outbound_size(self) -> int:
        return self._outbound.qsize()


_queue: MessageQueue | None = None


def get_message_queue() -> MessageQueue:
    global _queue
    if _queue is None:
        _queue = MessageQueue()
    return _queue
